src/gc/collector.cpp

```diff
@@ -283,30 +299,47 @@ void GCVisitor::visitPotentialRange(void* const* start, void* const* end) {
     }
 }
 
-static int ncollections = 0;
+static __attribute__((always_inline)) void visitByGCKind(void* p, GCVisitor& visitor) {
+    assert(((intptr_t)p) % 8 == 0);
 
-void markPhase() {
-#ifndef NVALGRIND
-    // Have valgrind close its eyes while we do the conservative stack and data scanning,
-    // since we'll be looking at potentially-uninitialized values:
-    VALGRIND_DISABLE_ERROR_REPORTING;
-#endif
+    GCAllocation* al = GCAllocation::fromUserData(p);
 
-#if TRACE_GC_MARKING
-#if 1 // separate log file per collection
-    char tracefn_buf[80];
-    snprintf(tracefn_buf, sizeof(tracefn_buf), "gc_trace_%03d.txt", ncollections);
-    trace_fp = fopen(tracefn_buf, "w");
-#else // overwrite previous log file with each collection
-    trace_fp = fopen("gc_trace.txt", "w");
-#endif
-#endif
-    GC_TRACE_LOG("Starting collection %d\n", ncollections);
-
-    GC_TRACE_LOG("Looking at roots\n");
-    TraceStack stack(roots);
-    GCVisitor visitor(&stack);
+    GCKind kind_id = al->kind_id;
+    if (kind_id == GCKind::UNTRACKED) {
+        // Nothing to do here.
+    } else if (kind_id == GCKind::CONSERVATIVE || kind_id == GCKind::CONSERVATIVE_PYTHON) {
+        uint32_t bytes = al->kind_data;
+        if (DEBUG >= 2) {
+            global_heap.assertSmallArenaContains(p, bytes);
```

Comment:
won't this cause usually cause assertions?  long objects do large conservative allocations, as will densemaps/vectors/etc.